### ISSUE TITLE
Argon2 seems like an unnecessarily slow choice of hash for CSRF tokens

### ISSUE THREAD

**Issue by reivilibre** [AscendingCreations/AxumCSRF#10]
Hi,
Not really a bug per se, but I was wondering if you had a specific reason to use Argon2?

I had it in mind that Argon2 was suited to password hashing and key derivation, but in this use case it seems overkill for generating and validating CSRF tokens.

The time it takes for Argon2 to generate a token is very noticeable in page loads (on my machine at least, etc etc :-)) and I worry this would scale poorly if it wasn't just me using the service.

(Sorry if this issue seems a bit rude — I should note that I'm grateful for you sharing your approach to anti-CSRF in Axum!)

**Comment by genusistimelord** [comment]
it was done to make it very difficult to reverse the hashing which could allow others to use the CSRF token to submit stuff they should not be able too.  If this is a issue We could add a feature to turn on a much faster hash but with less security. As you do not want others to be able to reuse this token in any way as that defeats the purpose of CSRF. 

To me its generally a Security First, Speed last in most situations.  For cases like where you must 100% need speed we can plan out something for that too.

**Comment by reivilibre** [comment]
Correct me if I'm wrong, but at a first approximation, for most intents and purposes you could just as well set the same value in the Cookie as it is in the Form, provided the value is securely random.
(My understanding is: the purpose of CSRF is to prevent another website tricking someone's browser into making a request on your site without the user's informed input / without going through a form on your website.
Because no mechanism exists that allows another website to set or read a Cookie on the user's browser for your domain, it can't construct a request with a matching value.)

---

(Bear with me on the bit, I imagine you probably know most or all of this already, but it's been a long time for me and I'm trying to work out the history here)

Turns out that is simplified and there's a catch: turns out it is actually somewhat possible for someone to write cookies to your domain in some circumstances: if another site on another subdomain of your domain is compromised (even just with XSS), or someone can MITM an unencrypted HTTP connection for any subdomain of your site, they can overwrite the cookies (even ones with the Secure flag set?).
— https://owasp.org/www-chapter-london/assets/slides/David_Johansson-Double_Defeat_of_Double-Submit_Cookie.pdf
N.B. It seems that cookies with a `__Host-` prefix cannot be overwritten from other subdomains in most modern browsers.
**→ This seems like a good recommendation for the default cookie name (tangential to the discussion but since I saw it here, seemed worth mentioning)**

OWASP claim that a more secure approach is the ['Signed Double Submit Cookie'](https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#signed-double-submit-cookie):

Assuming a server-side `secret` is already chosen,

1. choose a random value
2. message = (the user's session ID, the random value)
3. csrf_token = hmac(sha256, secret, message) || message
4. store csrf_token in a cookie
5. put message in the HTML form

The properties here:

- an attacker can't construct their own csrf_token of choice because they don't have the secret for the HMAC.
- the CSRF token must be dependent on some sort of session ID for the user (and I think you must check this matches when verifying it)
   - Oddly, they only 'recommend' this.
   - But to me this seems like it is a requirement for this system to be any more secure than the 'naive' solution with plaintext cookie = form value...
   - I say that because the attacker can of course just make a request to your website to obtain their own (csrf_token, message) pair and then they can still implant it in the victim's cookies / use the message in the forged form.
   - If the attacker *can't* implant the cookies in this case, then they couldn't implant the cookies in the naive case either!
   - If you see what I'm missing, please let me know as this is bugging me.

---

The implementation in axum_csrf seems like this from what I can tell:

- csrf_token = random value
- cookie = encrypt(csrf_token) (but also sign: it's authenticated encryption)
- form_token = argon2-hash(csrf_token, application-wide salt)

I think this is also vulnerable to an attacker just generating their own cookie and form_token in their own browser and then injecting the cookie into a victim and using their corresponding form_token in a forged request.

I am struggling to see how the hash strength is providing so much security here. (I also note that the form_token is hashed on every page load when the token is needed, even though it appears that the hash value should be the same each time.)

With that said I should be in bed (should have been a while ago) so I'm hoping I can spot something tomorrow.

**Comment by genusistimelord** [comment]
Only a direct VPN can protect you from a MITM attack I hate to say. CSRF is mostly for ease droppers and form re-submitter. If you have a MITM attack occurring it means your entire internet connection is already screwed.

Making them more secure can help prevent people from figuring out your backend keys. this is why you should never send out plain text form_tokens that have not been hashed. As they can figure out what the encryption keys are for the secure cookie. Which also lets them recreate Secure cookies and even recreate a proper CSRF passable key. 

The only way to prevent the Above would be to link the CSRF Token to a Session in which you add a third check to the keys to ensure neither one got changed and that everything matches 100% across the board. As if they changed the keys and it 100% worked but it fails the check against the key originally sent out then you have a issue and you need to swap your keys out. But yeah leaving the Hash in a less more vulnerable hash can allow them to reverse engineer the encryption, which is what we are trying to prevent.

oh and here is that example code which is in the readme that shows how to use the session to store the original key for the next check. Which prevents replay attacks and prevents people figuring out the keys and trying to make new CSRF tokens. 

```rust
async fn check_key(token: CsrfToken, session: Session<SessionPgPool>, Form(payload): Form<Keys>,) -> &'static str {
    let authenticity_token: String = session.get("authenticity_token").await.unwrap_or_default();

    if let Err(_) = token.verify(&payload.authenticity_token) {
        "Token is invalid"
    } else if let Err(_) = token.verify(&authenticity_token) {
        "Modification of both Cookie/token OR a replay attack occured"
    } else {
        // we remove it to only allow one post per generated token.
        session.remove("authenticity_token").await;
        "Token is Valid lets do stuff!"
    }
}
```

But yeah in a MITM attack nothing can protect you other than a VPN directly to the Server you are trying to access.

**Comment by reivilibre** [comment]
I think I should maybe clarify that by MITM, I mean here that it could be another subdomain of yours has its plaintext traffic intercepted (not HTTPS). HSTS might be partially useful to prevent this.

I think there's something I'm missing though:

Ignoring the session example, the CSRF tokens are currently not single use, nor are they somehow bound to a single user. This means that an attacker could just log on to your website and generate a CSRF form token and cookie.
At this point they could use one of the cookie fixation attacks to give the cookie to a victim user and then use the form token in a forged request.

If they *couldn't* do this, then there's nothing lost by the server just generating a random string and giving it to the client's cookie and putting it in the form. Because an attacker can't sniff it out anyway.
(the only flaw I can see is that a BREACH/CRIME compression attack leaves a potential for a side channel analysis method here if you use HTTP or TLS compression, so you might want at least some minimal hashing to avoid that if that's in use)

Do you see what I'm getting at, or am I completely insane? :-) Either the attacker can give a cookie to a victim, or they can't. Whether they choose their own value or ask your server to do it for them seems like it doesn't matter.

*The only thing that seems to protect you fully from this, as far as I can see, is if you tie the CSRF token to the user's session in some way. At that point you want the cookie to be at least signed so you know your server is the one that generated it **for that specific user**.*

As a note on 'leaving the Hash in a less more vulnerable hash can allow them to reverse engineer the encryption, which is what we are trying to prevent': I don't believe this is correct, or at least not if you choose the correct encryption scheme.
If you give an attacker some plaintext M and the encrypted version `Enc(K, M)` which was encrypted with key `K`, they should not be able to recover the key `K`. (even if you give them multiple different plaintext-ciphertext pairs)
(Some encryption schemes use a 'nonce' as part of the mechanism to ensure this, so long as the nonce is not reused.)
If you are interested in the formal definitions here, it is somewhat worth looking at the IND-CCA1 and IND-CCA2 properties. They don't say exactly the same thing but they actually state a stronger  property I believe.

**Comment by genusistimelord** [comment]
I will switch the default hash over to HMAC but I will also allow the Argon hash to stay for those who would want it regardless of how useless someone might think this is. 
I will set a regenerated Message that is appended to the Secure Cookie which is used to hash the Token this can work for both Argon and HMAC. 

The above would be to avoid me requiring someone to use my Session library but if you think this is still needed to be enforced for better Security I can integrate it in and enforce usage of this. As you know developers these days cant be trusted to do some work themselves. If i integrate the Session lib then.


Still switch to hmac with argon support if wanted.
Still regenerate the Message but to not place it into the Secure cookie and instead store it server side within the session (which is more secure) 

I do need to be careful to not do to much checking and setting inside of the service as I want the end user to have the capability to handle the bad request on their own, Which means in the Verify part it will need to handle the removal of the CSRF from the session.

**Comment by reivilibre** [comment]
I spoke to a friend, who does security for a job, about CSRF (not this specific library, but in general CSRF and how the OWASP document seems to be ... misleading).

One of them remarked that you have to be careful how you generate CSRF tokens because he once took down a site by repetitively requesting CSRF tokens as they were so CPU intensive (so it's an easy DoS vector). This made me think of the current case with Argon2 here :^).

One noteworthy remark is that:
- if you are trying to protect against cross-site request forgery because your requests are authenticated by a cookie, modern browsers provide a solution:
    - set the Cookie to have a prefix name of `__Host-`,
    - make it a `Secure` and `SameSite=Lax` cookie (and consider `HttpOnly` depending on the circumstances)
        - `SameSite=Lax` means that the cookie won't be sent if the request originates from a different site (e.g. in a `<form>` in an `<iframe>` — the classic CSRF scenario)
            - `SameSite=Strict` is likely to be too strict as it means that someone clicking a link from another site won't be logged in when they land at your site, even for a GET request
            - `SameSite=Lax` protects POST requests so is fine
        - the `__Host-` prefix means it can't be set by other (sub)domains that would normally be able to set cookies for your (sub)domain: that means that even a compromised site on one of the related (sub)domains won't be able to cause a cookie fixation attack like what https://owasp.org/www-chapter-london/assets/slides/David_Johansson-Double_Defeat_of_Double-Submit_Cookie.pdf mentions

Browsers which don't support `SameSite=Lax` and `__Host-` properly are rare, but old browsers don't support it and maybe some niche or experimental browsers may not right now either.

So at that point there's a question about what you want to do:

- ignore them? (I personally don't like this idea, even if they are really in the minority)
- use a basic CSRF scheme? (e.g. by just using a plain random CSRF token, or something a little more clever to prevent the potential for it being extracted using a BREACH/CRIME attack)
    - this is still safe unless one of your subdomains gets compromised/is vulnerable to XSS.
    - but will an attacker really bother to mount a CSRF attack when they can only affect people on very old or niche browsers? I would think unlikely.
- use a fully-secure CSRF scheme that binds the CSRF token to the session (e.g. by storing the CSRF token in session storage, or by using a HMAC of a string including the session ID)?
   - obvious pain is that you have to integrate a session system with your CSRF code
   - (n.b. I checked with my security friend and he told me that his experience with anti-CSRF systems is that they are usually session-bound, yes — if it's not then he agrees that the signed cookie solution does not provide more security than the naive cookie solution)

I would think different people will have different answers to what they want.
And you are correct in saying that people won't want to necessarily adopt your session system, so I don't know what the right answer here for you is, but my recommendations that I *can* feel comfortable giving:

1. use something other than Argon2, because it currently is more of a risk / DoS vector than the security it provides in the circumstance
2. the CSRF cookie should start with a `__Host-` prefix by default → currently it is `Csrf_Token`, so e.g. `__Host-Csrf_Token` would be advisable
3. the CSRF cookie should have `SameSite=Lax` by default: it already does :)
4. The session cookie should have `SameSite=Lax`. `__Host-` is also advisable to prevent session fixation.

(4) is out of scope of this library, but worth bearing in mind for developers.

With (2) and (3), the CSRF cookie can't be stolen by, or overwritten by, compromised subdomains, in modern browsers. This means you have full security for modern browsers.

(1) is to make it so I can't bring down your server just by loading a few pages in parallel...

**Comment by reivilibre** [comment]
(by the way, I would also suggest `__Host-` prefixing the session cookie name in your `axum_session` library. That would help protect you a bit from session fixation using the same mechanism here!)

**Comment by genusistimelord** [comment]
can we discuss this further on my discord channel? as this would be easier to discuss there are least. 
https://discord.gg/gVXNDwpS3Z

**Comment by genusistimelord** [comment]
so apparently I can not enforce __HOST- I will have to make it Optional as if I enforce it then regular http pages can not get cookies. as it requires Secure to be Set and the page be from HTTPS and the path to be /. Personally i should not need to add anything here since the end user can Create whatever cookie name they want the cookie to be already.

**Comment by reivilibre** [comment]
Hi, sorry for going away for a while. I don't have Discord so won't be able to join you there.

I think there is a lot of value in the defaults being secure: if we want CSRF protection to be worthwhile then we should be using HTTPS anyway. I would hazard a guess that most people are using HTTPS if they care about this type of thing, anyway.
The main exception I can think of is during development on `localhost`, but browsers have exceptions for localhost for most of these things (I don't know specifically about `__Host-`-prefix but for the `Secure` attribute they do at least).

**Comment by genusistimelord** [comment]
@reivilibre I have updated the main branch with Hmac to replace Argon2. I also added a bool to enable __HOST- appending the the cookie name. This should resolve your issues. For more Security I do suggest using https://crates.io/crates/axum_session to store a Copy of the key per Session to ensure the Key did not get changed by anyone else for any reason. 

I did think about adding a part that would Auto remove the cookie when called at the end of Post But the issue is if the perpetrator did indeed have a Copy of the cookies they could just do a replay attack anyways. So its best to set a Session up to avoid this.
